### PR TITLE
Feature/fallback hosts

### PR DESCRIPTION
**Pull request by ChrisWiegman**
This gives the option to have one or more fallback servers should the primary ES host fail. Fallback servers can be specified in a global array $eb_backup_hosts. Upon failure of the primary host the backup hosts will be checked sequentially until a good host is found or an error will be returned. Good hosts are cached for an hour before the process is checked again.

### CODE REVIEW

**Comment by tlovett1** [comment]
@ChrisWiegman can we add tests?

**Comment by ChrisWiegman** [comment]
Good point. I'm usually told not to add tests with Voltron work. Let me see about doing something there.

**Comment by ChrisWiegman** [comment]
Added some tests to verify the ep_get_host function. Tests ran successfully. Note this is a rather strange scenario as we're defining the primary EP_HOST in bootstrap by necessity however this would normally make testing the function difficult. To overcome I've introduced the use of an extra constant that forces the bypass of the defined host in favor of the backup array. This should be CRed before further push.

**Comment by tlovett1** [comment]
@ChrisWiegman seeing two things that might be issues
1. `ep_get_hosts` only seems to be used for some of the API methods. For example, it doesn't seem to be used in search?
1. I think we can be a bit more clever with the number of requests sent to the ES server. Instead of checking if a host is available in `ep_get_hosts`, why not just check the response code of `wp_remote_request`? If it 404's, try the next host. This is what I meant by a `wp_remote_request` wrapper.

**Comment by ChrisWiegman** [comment]
Gotcha. Still getting familiar so it is highly possible I missed some. I'll replace the wp_http calls instead which will help. I also realized this morning that it wouldn't automatically fail over for up to an hour when I looked again this morning. Let me clean this up and I'll resubmit.

File: classes/class-ep-api.php [hunk 1]
```diff
@@ -1236,6 +1241,44 @@ public function elasticsearch_alive() {
...
+	 * @since 1.5
+	 *
+	 * @param string $url  Site URL to retrieve.
+	 * @param array  $args Optional. Request arguments. Default empty array.
+	 *
+	 * @return WP_Error|array The response or WP_Error on failure.
+	 */
+	public function remote_request( $url, $args = array() ) {
+
+		$host = ep_get_host();
```
Comment (tlovett1) [inline classes/class-ep-api.php]: Where does this get used?

File: classes/class-ep-api.php [hunk 2]
```diff
@@ -1236,6 +1241,44 @@ public function elasticsearch_alive() {
...
+	 * (if present) should the primary EP host fail.
+	 *
+	 * @since 1.5
+	 *
+	 * @param string $url  Site URL to retrieve.
+	 * @param array  $args Optional. Request arguments. Default empty array.
+	 *
+	 * @return WP_Error|array The response or WP_Error on failure.
+	 */
+	public function remote_request( $url, $args = array() ) {
```
Comment (tlovett1) [inline classes/class-ep-api.php]: Seems like it makes sense to use path instead of url. Instead of passing the full index URL to each remote_request call, calculate the full URL in here.

File: classes/class-ep-api.php (hunk 1 above)
Comment (ChrisWiegman) [inline classes/class-ep-api.php]: In an attempt to maintain consistency with existing code there is an accessor function, ep_remote_request, that makes use of it.

File: classes/class-ep-api.php (hunk 2 above)
Comment (ChrisWiegman) [inline classes/class-ep-api.php]: I could do that, the existing code seemed to have a couple different methods for calculating the $url, when first written I actually was going to extract it however in order to make it more compatible with what is there I elected to take this approach.

File: classes/class-ep-api.php [hunk 3]
```diff
@@ -387,6 +409,10 @@ public function prepare_post( $post_id ) {
 		$post_date_gmt = $post->post_date_gmt;
 		$post_modified = $post->post_modified;
 		$post_modified_gmt = $post->post_modified_gmt;
+		$comment_count = absint( $post->comment_count );
```
Comment (tlovett1) [inline classes/class-ep-api.php]: This should be in a separate PR.

File: classes/class-ep-api.php [hunk 4]
```diff
@@ -543,11 +573,23 @@ public function prepare_meta( $post ) {
 	 */
 	public function delete_index( $index_name = null ) {
 
-		$index_url = ep_get_index_url( $index_name );
+		if ( null === $index_name ) {
+
+			$index = ep_get_index_name();
+
+		} elseif ( is_array( $index_name ) ) {
```
Comment (tlovett1) [inline classes/class-ep-api.php]: Array support should be in a separate PR.

File: classes/class-ep-api.php [hunk 5]
```diff
@@ -568,11 +610,24 @@ public function delete_index( $index_name = null ) {
...
 	 */
 	public function index_exists( $index_name = null ) {
-		$index_url = ep_get_index_url( $index_name );
 
-		$request_args = array( 'headers' => $this->format_request_headers() );
+		if ( null === $index_name ) {
+
+			$index = ep_get_index_name();
+
+		} elseif ( is_array( $index_name ) ) {
```
Comment (tlovett1) [inline classes/class-ep-api.php]: Array support should be in a separate PR

File: classes/class-ep-api.php [hunk 6]
```diff
@@ -636,13 +691,17 @@ public function format_args( $args ) {
...
-			$formatted_args['sort'] = array(
+			$default_sort = array(
 				array(
 					'_score' => array(
 						'order' => $order,
 					),
 				),
 			);
+
+			$default_sort = apply_filters( 'ep_set_default_sort', $default_sort, $order );
```
Comment (tlovett1) [inline classes/class-ep-api.php]: Wondering if maybe you should rebase this PR?

File: classes/class-ep-api.php [hunk 7]
```diff
@@ -877,6 +936,17 @@ public function format_args( $args ) {
 							}
 
 							break;
+						case 'like':
```
Comment (tlovett1) [inline classes/class-ep-api.php]: Wondering if maybe you should rebase this PR?

File: classes/class-ep-api.php [hunk 8]
```diff
@@ -1236,6 +1316,61 @@ public function elasticsearch_alive() {
...
+		//If we have a failure we'll try it again with a backup host
+		if ( false === $request || is_wp_error( $request ) || ( isset( $request['response']['code'] ) && 200 !== $request['response']['code'] ) ) {
+
+			$host = ep_get_host( true, $use_backups );
+
+			if ( is_wp_error( $host ) ) {
+				return $host;
+			}
+
+			return wp_remote_request( esc_url( trailingslashit( $host ) . $path ), $args );
```
Comment (tlovett1) [inline classes/class-ep-api.php]: Seems like this only supports one backup host. I was thinking more like a loop of all backup hosts with a filter to limit the amount of attempts.

File: classes/class-ep-config.php [hunk 9]
```diff
@@ -19,6 +19,67 @@ public static function factory() {
...
+	public function get_ep_host( $force = false, $use_only_backups = false ) {
+
+		global $ep_backup_host;
+
+		// Delete the transient if we want to force a new good host lookup
+		if ( true === $force ) {
+			delete_transient( 'ep_last_good_host' );
+		}
+
+		$last_good_host = get_transient( 'ep_last_good_host' );
```
Comment (tlovett1) [inline classes/class-ep-config.php]: Remember this is sometimes intended to run site-wide. Might make more sense to use site transients.

File: classes/class-ep-api.php (hunk 3 above)
Comment (ChrisWiegman) [inline classes/class-ep-api.php]: The odd thing about this one is it is present before my changes (see https://github.com/10up/ElasticPress/blob/aafc3999487598347335b2dc38b9979b8c7eff74/classes/class-ep-api.php#L394) so as to why it keeps showing as a newline in this changeset is interesting. I'll see if I can get that one out in the rebase.

File: classes/class-ep-api.php (hunk 4 above)
Comment (ChrisWiegman) [inline classes/class-ep-api.php]: Fixed

File: classes/class-ep-config.php (hunk 9 above)
Comment (ChrisWiegman) [inline classes/class-ep-config.php]: Excellent point

File: classes/class-ep-api.php (hunk 7 above)
Comment (ChrisWiegman) [inline classes/class-ep-api.php]: Rebasing complete. (GH really needs a better way to check these off before a push. Push coming shortly).

File: classes/class-ep-api.php (hunk 6 above)
Comment (ChrisWiegman) [inline classes/class-ep-api.php]: Rebasing complete

File: classes/class-ep-api.php (hunk 5 above)
Comment (ChrisWiegman) [inline classes/class-ep-api.php]: Removed for now. Will do a new PR after this is merged

File: classes/class-ep-api.php (hunk 8 above)
Comment (ChrisWiegman) [inline classes/class-ep-api.php]: It actually does the loop, just not here. The first time it calls ep_get_host it will use the transient (if available). Then, the host is verified and if it fails it goes back to ep_get_host with the force param set to true. It is here that it will loop through all the available hosts until a success is found. This just makes it slightly more efficient in the case of a cached host.

**Comment by tlovett1** [comment]
This looks solid. I'm merging. @allan23 mind spending a little time QAing this?
